src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs

```diff
@@ -357,12 +357,17 @@ static IEnumerable<(string Pattern, string Input, RegexOptions Options, int Begi
                     yield return ("(?(dog2))", "dog2", RegexOptions.None, 0, 4, true, string.Empty);
                     yield return ("(?(a:b))", "a", RegexOptions.None, 0, 1, true, string.Empty);
                     yield return ("(?(a:))", "a", RegexOptions.None, 0, 1, true, string.Empty);
+                    yield return ("(?(cat)cat|dog)", "cat", RegexOptions.None, 0, 3, true, "cat");
+                    yield return ("(?((?=cat))cat|dog)", "cat", RegexOptions.None, 0, 3, true, "cat");
                     yield return ("(?(cat)|dog)", "cat", RegexOptions.None, 0, 3, true, string.Empty);
                     yield return ("(?(cat)|dog)", "catdog", RegexOptions.None, 0, 6, true, string.Empty);
                     yield return ("(?(cat)|dog)", "oof", RegexOptions.None, 0, 3, false, string.Empty);
                     yield return ("(?(cat)dog1|dog2)", "catdog1", RegexOptions.None, 0, 7, false, string.Empty);
                     yield return ("(?(cat)dog1|dog2)", "catdog2", RegexOptions.None, 0, 7, true, "dog2");
                     yield return ("(?(cat)dog1|dog2)", "catdog1dog2", RegexOptions.None, 0, 11, true, "dog2");
+                    yield return (@"(?(\w+)\w+)dog", "catdog", RegexOptions.None, 0, 6, true, "catdog");
```

Comment:
I'm having trouble parsing `(?(\w+)\w+)dog`

I figured this was equivalent to `(?(\w+)\w+|)dog` but that has a 2nd match of empty string at the end of the text. So it's a conditional, but without an else? The docs hint at this but aren't explicit.

Reply:
`(?(\w+)\w+)dog` is equivalent to `(?(\w+)\w+|)dog`; the former lacks an else block, and the latter has an empty else.  Can you clarify "but that has a 2nd match of empty string at the end of the text"?

Reply:
`Regex.Matches("abcd", @"(?(\w+)\w+)")` returns one match, `Regex.Matches("abcd", @"(?(\w+)\w+|)")` returns two.

Reply:
That's definitely not what I have implemented now.  Do you have an explanation for why that makes sense? Seems like a bug.

Reply:
I'm not at my machine... what do you get for `Regex.Matches("", @"(?(\w+)\w+)")`

Reply:
Oh, maybe the semantics are a failure to match the condition with no else branch should be interpreted as match nothing rather than match empty? I implemented the latter, but I see the argument for the former and can change to that.

Reply:
> Oh, maybe the semantics are a failure to match the condition with no else branch should be interpreted as match nothing rather than match empty? I implemented the latter, but I see the argument for the former and can change to that.

Nope.  If that were the case, this wouldn't match, but it does:
```C#
Regex.IsMatch("", @"(?(\w+)\w*)")
```
So, I'm back to "Do you have an explanation for why that makes sense? Seems like a bug."

Reply:
I think match nothing makes more sense to me and is also more useful. 

It's also arguably analogous to the difference `a` vs `a|` (an alternation with an empty alternative).

Having said that I experimented with PCRE. Perl like most (?all) other engines, only allows assertions and capture references in its condition so I used an assertion. Playing then with something like `(?(?<!x)a|)` against `xaxba` [one gets ](https://regex101.com/r/iYn3cx/1) 
![image](https://user-images.githubusercontent.com/6385855/148708058-08281568-9259-4072-9c0b-f5af208649bf.png)
and without the `|` the same:
![image](https://user-images.githubusercontent.com/6385855/148708084-8bd6c0b0-cb91-4664-9c8f-de91afb6244f.png)

What proportion of your corpus use conditionals (perhaps you could put the stats on a gist?) and do any have missing negative clauses? This seems like another dusty corner that maybe wasn't carefully though through. For example there's no simple way I can see to use a literal single digit as the condition. I tried `(?([1])b|c)` but it is treated literally itself.

Reply:
> I think match nothing makes more sense to me

I see arguments for both.

>  and is also more useful.

They're equally useful.  You can implement "empty" as:
```
(?(abc)def|)
```
and you can implement "nothing" as:
```
(?(abc)def|(?!))
```
so it's just a question of which it's supposed to be.

> What proportion of your corpus use conditionals (perhaps you could put the stats on a gist?) and do any have missing negative clauses?

101 backreference conditionals and 5 expression conditionals do not have an else block.  Most of the ones I sampled are of the same form as what's used in the docs at https://docs.microsoft.com/en-us/dotnet/standard/base-types/grouping-constructs-in-regular-expressions#balancing_group_definition, e.g. `(?(Open)(?!))`.  That construct assumes the "match empty" semantics and makes no sense under the "match nothing" semantics, as it's saying "if the Open capture group matched, then fail"... if the lack of an else also implicitly failed, then this whole expression would always fail.

Reply:
"match empty" is required by pretty much all the real examples I found.  It's also what multiple tutorials claim the behavior should be.  I've gone with that.